### ISSUE TITLE
Need to examine 2D data pixel sizes

### ISSUE THREAD

**Issue by butlerpd** [SasView/sasview#1544]
This issue has been raised before, but Mathieu made me start wondering about this in more detail when he noted some strangeness with 2D data which was in the *.DAT format.  He traced it down to the fact that the .DAT reader has the following lines:
```Python
243        # calculate the number of pixels in the each axes>
244        npix_y = math.floor(math.sqrt(len(data)))
245        npix_x = math.floor(len(data) / npix_y)
246
247        # calculate the size of bins
248        xstep = x_size / (npix_x - 1)
249        ystep = y_size / (npix_y - 1)
```
found at:
https://github.com/SasView/sasview/blob/626b30ba86299cee55c87ee70a1a99b55ad027bf/src/sas/sascalc/dataloader/readers/red2d_reader.py#L243

In other words it is calculating the pixel size and to do so is assuming a square detector with even rows and columns... which is not necessarily true.

In principle "pixel size" is not a valid concept for reduced data which has a normalized intensity at Q over a resolution in sigmaQx and sigmaQy.  However when it comes time to graphically represent, it is not clear how to best use that information.  I would argue a bit of research probably needs to be done (don't think it is a 'solved problem') but that in fact that is exactly what should be used to calculate the graphical representation.

In the meantime it seems that the plotting is, by necessity, using the concept of pixels. In principle the most generally correct answer would be to calculate on the fly by assigning half the space between two adjacent Q points to each Q. However I can see that with a large number of Q points which are not on a grid, that could be problematic.

This raises a few points
1) How is the pixel size obtained now? Does the NXcanSAS format contain something that can be used?
2) How does the pixel size get stored in the SasView data2D object? does it allow for non square pixels? does it allow for different pixel sized for each Q point (to be general it should, but ...).
3) Where/how does that (those) pixel size(s) get used.
4) if the Q data are not on a grid what happens?

**Comment by krzywon** [comment]
This is likely only an issue with plotting of data and is because the reader is assuming a square detector. For fitting purposes, sasmodels only uses Qx, Qy, I and their resolutions. Sizes and shapes are ignored. Plotting is not using pixel size in any way, only the number of x and y pixels.

All data objects are stored as 1D arrays. The Data2D object has x_bins and y_bins variables to define the plotting shape. In 4.x, this was fixed by Lewis (#754). Looking at [Plotter2D](https://github.com/SasView/sasview/blob/9703e10275fe7357108e4b6fd42fa834a9c8b742/src/sas/qtgui/Plotting/Plotter2D.py#L470) in the ESS_GUI branch, it appears 5.x does something similar.

Calculating the shape of the detector based on the Qx and Qy data is a pretty good suggestion and would only need to be applied to the DAT reader when loading the data in. Other readers can already determine this information based on the data file (see point 4 below). Qx and Qy values should be approximately equal in each row and column, so the loader could use that to reconstruct the shape.

Answering your questions:
1. The [NXcanSAS ](https://manual.nexusformat.org/classes/applications/NXcanSAS.html) format has that information in the NXdetector object as x_pixel_size and y_pixel_size. The [canSAS 1D](http://www.cansas.org/formats/canSAS1d/1.1/doc/element_SASdetector.html) format also has these elements as well as pixel_size.z. I can't speak for other formats.
2. [DataInfo()](https://github.com/SasView/sasview/blob/9703e10275fe7357108e4b6fd42fa834a9c8b742/src/sas/sascalc/dataloader/data_info.py#L530) instances hold a list of [Detector()](https://github.com/SasView/sasview/blob/9703e10275fe7357108e4b6fd42fa834a9c8b742/src/sas/sascalc/dataloader/data_info.py#L179) instances. The [pixel_size](https://github.com/SasView/sasview/blob/9703e10275fe7357108e4b6fd42fa834a9c8b742/src/sas/sascalc/dataloader/data_info.py#L201) variable is in each Detector instance. Pixels are (wrongly) assumed to be square in shape and this should be fixed. Should this be a separate ticket or keep it as part of this?
3. Pixel size isn't used for any purpose within SasView. It is considered meta-data. Plotting should use the x_bins and y_bins variables to determine how to plot the data.
4. NXcanSAS 2D data are stored as and loaded in as 2D array-like objects, so the point is moot. The shape of the data is already known. The only 2D data recognized by SasView that isn't stored as a grid is the DAT reader.

I am already working on a large-ish change to the reader system in the [ESS_GUI_loader_refactor](https://github.com/SasView/sasview/tree/ESS_GUI_loader_refactor) branch, so I can take this as a part of that work.

**Comment by krzywon** [comment]
> Pixels are (wrongly) assumed to be square in shape and this should be fixed. Should this be a separate ticket or keep it as part of this?

**Correction:** X, Y, and Z dimensions can be defined for the pixel size.

data_info.py, line 214: `        self.pixel_size = Vector()`
